Analytic jacobian

hwpang opened the pull request:
This PR enables calculating analytic jacobian for individual domains, with the exception that the partials with respect to thermodynamic variables are calculated using forward automatic differentiation. See #55. 

This PR also changes the p in ConstantTVDomain to store non-diffusion-limited forward reaction rates (kfsnondiff), instead of storing diffusion-limited (kfs). This helps to obtain correct diffusion-limited reverse reaction rates (krevs).

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1577,6 +1577,29 @@ end
+            flow = inter.F(t)
+            jac[domain.indexes[1]:domain.indexes[2],domain.indexes[1]:domain.indexes[2]] .-= flow/N*Matrix(1.0I,domain.indexes[2]-domain.indexes[1]+1,domain.indexes[2]-domain.indexes[1]+1)
+            for i in domain:indexes[1]:domain.indexes[2]
+                jac[domain.indexes[1]:domain.indexes[2],i] .-= -flow*ns./N^2
```
This might show up in the numerical jacobians because we're calculating N by summing the ns instead of through N = P*V/(R*T) as we should be.

hwpang replied:
Thanks for the review! I changed the definition of N to PV/(RT) instead of sum(ns) for the calcthermo of individual domains.

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1599,6 +1599,49 @@ function jacobiany!(jac::Q,y::U,p::W,t::Z,domain::D,interfaces::Q3,colorvec::Q2=
+    ns,cs,T,P,V,C,N,mu,kfs,krevs,Hs,Us,Gs,diffs,Cvave = calcthermo(domain,y,t,p)
+    jacobiany_ns!(jac,y,p,t,domain,[],cs,kfs,krevs,nothing)
+    
+    cpdivR,hdivRT1,sdivR = calcHSCpdless(domain.phase.vecthermo,T)
```
It should be easy for us to add cpdivR from the calcthermo functions with virtually no computational cost (since the vector is already created in calcthermo for these domains).

hwpang replied:
I modified all the calcthermo functions to output cpdivR when T is not constant or parametrized. For isothermal or parametrized T Domain, it outputs Array{Float64,1}() instead.

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1577,6 +1577,29 @@ end
+    end
+    
+    for inter in interfaces
+        if isa(inter,Outlet) && domain == inter.domain
```
Aren't we missing the contribution of the outlet to the reactor volume here?

hwpang replied:
> Aren't we missing the contribution of the outlet to the reactor volume here?

Since we use molar flow to specify the outlet, the contribution from the outlet flow is dV/dt -= inter.F(t)*R*T/P, so d/dni dV/dt -= 0

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1577,6 +1577,29 @@ end
+    ns,cs,T,P,V,C,N,mu,kfs,krevs,Hs,Us,Gs,diffs,Cvave = calcthermo(domain,y,t,p)
+    jacobiany_ns!(jac,y,p,t,domain,[],cs,kfs,krevs,nothing)
+    for i in domain.indexes[1]:domain.indexes[2]
+        jac[domain.indexes[3],i] = sum(jac[domain.indexes[1]:domain.indexes[2],i])*R*T/P
```
an @views on the jac slice might help avoid allocating a new vector corresponding to jac[domain.indexes[1]:domain.indexes[2],i] each iteratioon.

hwpang replied:
Thanks for the suggestion! Added!

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1599,6 +1599,49 @@ function jacobiany!(jac::Q,y::U,p::W,t::Z,domain::D,interfaces::Q3,colorvec::Q2=
+    dydt = zeros(size(y))
+    dydtreactor!(dydt,y,t,domain,interfaces,;p=p,sensitivity=false)
+    
+    dCvavedn = zeros(domain.indexes[2]-domain.indexes[1]+1)
```
It'd be nice if we could avoid allocating this vector. I think it might be okay to just recalculate dCvavedn as needed. It's not that expensive to recalculate and in most cases I expect this won't have an outlet and it won't matter at all.

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1790,6 +1790,23 @@ function jacobiany!(jac::Q,y::U,p::W,t::Z,domain::D,interfaces::Q3,colorvec::Q2=
     return jac
 end
 
+function jacobiany!(jac::Q,y::U,p::W,t::Z,domain::D,interfaces::Q3,colorvec::Q2=nothing) where {Q3<:AbstractArray,Q2,Q<:AbstractArray,U<:AbstractArray,W,Z<:Real,D<:AbstractDomain}
```
What Domains is this function intended for? Is this capable of generalizing to arbitrary new domains someone might write into RMS?

hwpang replied:
This is intended for ConstantTVDomain and ConstantTADomain. I will use Union{ConstantTVDomain, ConstantTADomain} instead.

hwpang commented on src/Domain.jl:
```diff
@@ -1322,12 +1322,18 @@ end
+        return ns,cs,d.T,P,d.V,C,N,d.mu,d.kfs,d.krevs,Array{Float64,1}(),Array{Float64,1}(),Array{Float64,1}(),Array{Float64,1}(),0.0
+    elseif nothermochg
+        d.kfsnondiff = p[length(d.phase.species)+1:length(d.phase.species)+length(d.phase.reactions)]
+        d.krevsnondiff = @fastmath d.kfsnondiff./getKcs(d.phase,d.T,d.Gs)
```
You're right! The krevsnondiff is redundant. It will be removed.

hwpang commented on src/Domain.jl:
```diff
@@ -1322,12 +1322,18 @@ end
+    elseif nothermochg
+        d.kfsnondiff = p[length(d.phase.species)+1:length(d.phase.species)+length(d.phase.reactions)]
+        d.krevsnondiff = @fastmath d.kfsnondiff./getKcs(d.phase,d.T,d.Gs)
+        d.kfs,d.krevs = getkfkrevs(;phase=d.phase,T=d.T,P=P,C=C,N=N,ns=ns,Gs=d.Gs,diffs=d.diffusivity,V=d.V,kfs=d.kfsnondiff)
```
Added!

hwpang commented on src/Domain.jl:
```diff
@@ -1343,8 +1349,10 @@ end
+    kfsnondiff = p[length(d.phase.species)+1:length(d.phase.species)+length(d.phase.reactions)]
+    kfs,krevs = getkfkrevs(;phase=d.phase,T=d.T,P=P,C=C,N=N,ns=ns,Gs=Gs,diffs=d.diffusivity,V=d.V,kfs=kfsnondiff)
+    kfs = convert(typeof(y),kfs)
+    krevs = convert(typeof(y),krevs)
```
Done!

hwpang commented on src/Reactor.jl:
```diff
@@ -148,6 +148,105 @@ function jacobianp(y::U,p::W,t::Z,domain::V,interfaces::Q3,colorvec::Q2=nothing)
+
+function jacobiany_ns!(jac::Q,y::U,p::W,t::Z,domain::D,interfaces::Q3,cs::Z1,kfs::Z2,krevs::Z3,colorvec::Q2=nothing) where {Z1,Z2,Z3,Q3<:AbstractArray,Q2,Q<:AbstractArray,U<:AbstractArray,W,Z<:Real,D<:AbstractDomain}    
+    @inline function _jacobian!(jac::AbstractArray,rxnarray::Array{Int64,2},cs::Array{Float64,1},k::Real,i::Int,rev::Bool=false)
+        j=0
```
I changed the function a little bit to avoid using j but still able to reuse this function!

hwpang commented on src/Reactor.jl:
```diff
@@ -247,6 +247,15 @@ function jacobiany_ns!(jac::Q,y::U,p::W,t::Z,domain::D,interfaces::Q3,cs::Z1,kfs
 end
 
+function jacobiany_therm!(jac::Q,y::U,p::W,t::Z,domain::D,interfaces::Q3,ind::I,x::F,colorvec::Q2=nothing) where {Q3<:AbstractArray,Q2,Q<:AbstractArray,U<:AbstractArray,W,Z<:Real,D<:AbstractDomain,I<:Int64,F<:Float64}
+    function _f(dy::X,x::T,y::Q,p::W,domain::D,interfaces::Q3,ind::T1) where {X,Q3<:AbstractArray,T<:Real,Q<:AbstractArray,D<:AbstractDomain,T1<:Integer,W}
```
Thanks for the review! The function is moved out!

hwpang commented on src/Domain.jl:
```diff
@@ -1343,8 +1349,10 @@ end
-    kfs = convert(typeof(y),p[length(d.phase.species)+1:length(d.phase.species)+length(d.phase.reactions)])
-    krevs = convert(typeof(y),getkfkrevs(;phase=d.phase,T=d.T,P=P,C=C,N=N,ns=ns,Gs=Gs,diffs=d.diffusivity,V=d.V,kfs=kfs)[2])
+    kfsnondiff = p[length(d.phase.species)+1:length(d.phase.species)+length(d.phase.reactions)]
+    kfs,krevs = getkfkrevs(;phase=d.phase,T=d.T,P=P,C=C,N=N,ns=ns,Gs=Gs,diffs=d.diffusivity,V=d.V,kfs=kfsnondiff)
```
Thanks for the comment!

I think the current code did the same thing as in ConstantTPDomain, i.e. recalculating krevs in each iterations. As you mentioned in our private conversation, recalculating the krevs in each iteration make the code more robust, in case that a user decides to change the parameters during integration using forward diff for jacobian.

I believe that the type change is handled by the next two lines of code that change kfs and krevs to type of y.

hwpang commented:
> Good job! I like the structure a lot!
> 
> 1. Lots of little optimizations
> 2. I've noticed we made a mistake in the sparse implementation...while we've reformulated the equations so that they can be calculated as sparse...we aren't actually calculating them in a sparse way...we should be calculating N = P_V/(R_T) instead of N = sum(n_i) which fixes this issue...but also changes the jacobians.
> 3. I'm not sure we're able to write an analytic jacobian for an arbitrary domain someone might write in RMS.
> 4. We should add tests for these jacobian calculators vs finite differences and probably auto diff as well from what you saw (to test auto diff vs FD too)
> 5. Since we're doing this foremost to speed up RMS jacobian calculations and simulations it's important that we confirm we're getting the speed up we expect from this (often we won't the first time and will need to optimize it a bit, this package can be handy for telling what's taking long: https://github.com/kimikage/ProfileSVG.jl) and measure roughly how much it is.

Thanks for the comment! @mjohnson541 
1. The code is modified as suggested!
2. I changed all the N to P*V/(R*T) except for the liquid phase domains
3. I was intending to write an analytic jacobian for ConstantTVDomain, ParametrizedTConstantVDomain and ConstantTADomain only. I changed the domain type to be the union of these three domains.
4. I added tests for these jacobian calculators vs. autodiff for all the domains. I didn't write a test for auto diff vs FD, because I realize that the situation I saw only appeared when I tested with ethane test mechanism. I didn't observe it in a more realistic model like the PME one you shared with me. Based on this, I think (1) it's possible that the ethane model wasn't very realistic, and (2) autodiff should work fine with more realistic application. Let me know if you think I should still add this test.
5. I tested the time required to simulate PME model of all the domains using this PR's jacobian calculator vs. forwarddiff. Generally we get pretty good speed up.
![image](https://user-images.githubusercontent.com/45482070/90450777-70d78c00-e0b8-11ea-8658-8eb9f0e91993.png)

hwpang commented on src/Reactor.jl:
```diff
@@ -247,6 +247,25 @@ end
     end
 end
 
+@inline function _calc_jacobiany_k_deriv!(jac::Array{Float64,2},xind::Int64,dkdx::Float64,reactantinds::Array{Int64,1},productinds::Array{Int64,1},cs::Array{Float64,1},V::Float64)
```
Thanks for the comment! I will add a docstring! So I'm using this function for both V and ni because they have the same deriv except for the dkdV or dkdni part. x can be either V or ni.

hwpang commented on src/Domain.jl:
```diff
@@ -1585,6 +1585,18 @@ end
+        _calc_jacobiany_k_deriv!(jac,spcind,dkdni,reactantinds,productinds,cs,V)
+        _calc_jacobiany_k_deriv!(jac,spcind,dkdni/Kc,productinds,reactantinds,cs,V)
+    end
+    @fastmath dkdV::Float64 = dkdCeff*(C-Ceff)/V
```
So Ceff = C + sum(ns(i)*val(i)))/V = (N + sum(ns(i)*val(i)))/V = (PV/RT + sum(ns(i)*val(i)))/V, and dCeffdV = -sum(ns(i)*val(i))/V^2 = (C-Ceff)/V

hwpang commented on src/Domain.jl:
```diff
@@ -1585,6 +1585,18 @@ end
 
+@inline function jacobiany_efficiency!(jac::Array{Float64,2},domain::Union{ConstantTPDomain,ConstantPDomain,ParametrizedTPDomain,ParametrizedPDomain},kinetics::AbstractFalloffRate,efficiencies::Dict{Int64,Float64},reactantinds::Array{Int64,1},productinds::Array{Int64,1},Vind::Int64,cs::Array{Float64,1},T::Float64,V::Float64,C::Float64,Ceff::Float64,Kc::Float64)
+    dkdCeff::Float64 = _calcdkdCeff(kinetics,T,Ceff)
+    for (spcind::Int64,efficiencyval::Float64) in efficiencies
```
This isn't necessary at all, I will remove them!

hwpang commented on src/Domain.jl:
```diff
@@ -1632,6 +1632,20 @@ end
 
+@inline function jacobiany_ns!(jac::AbstractArray,domain::Union{ConstantPDomain,ParametrizedPDomain},rxnarray::Array{Int64,2},efficiencyinds::Array{I,1},cs::Array{Float64,1},kfs::Array{Float64,1},krevs::Array{Float64,1},T::Float64,V::Float64,C::Float64) where {I<:Integer}
+    @simd for rxnind = 1:size(rxnarray)[2]
+        _calc_jacobiany_ns_deriv_ns!(jac,rxnarray[1:3,rxnind],rxnarray[4:6,rxnind],cs,kfs[rxnind])
```
So I chose to pass the rxnarray separately here so that I can reuse _calc_jacobiany_ns_deriv_ns! for both reactant direction and product direction, and it also helps to make  _calc_jacobiany_ns_deriv_ns!  more readable (Instead of having lots of rxnarray slicing inside the function, we only have the slicing at the function input). I will add **@views** !

hwpang replied:
I tested two version of jacobiany_ns!, one uses slicing with **@views**, and the other avoids all these slicing, and I found that the latter one is slightly faster (1 second) for PME mechanism ConstantTPDomain simulation. I will change all of the relevant functions to minimize the usage of slicing (even though using slicing gives clear and simpler code).

hwpang commented on src/Reactor.jl:
```diff
@@ -16,11 +16,7 @@ function Reactor(domain::T,y0::Array{W,1},tspan::Tuple,interfaces::Z=[];p::X=Dif
     dydt(dy::X,y::T,p::V,t::Q) where {X,T,Q<:Real,V} = dydtreactor!(dy,y,t,domain,interfaces,p=p)
     jacy!(J::Q2,y::T,p::V,t::Q) where {Q2,T,Q<:Real,V} = jacobiany!(J,y,p,t,domain,interfaces,nothing)
     jacp!(J::Q2,y::T,p::V,t::Q) where {Q2,T,Q<:Real,V} = jacobianp!(J,y,p,t,domain,interfaces,nothing)
-    if domain isa Union{ConstantTPDomain,ConstantTVDomain}
```
Good catch! I will change it such that it uses analytic jacobian for known domains and uses forwarddiff jacobian for unknown domains.

hwpang commented:
> Awesome! We're quite close! I think there are a number of places here where we aren't properly taking advantage of @inbounds and @views. A couple naming things and one design thing related to handling new domains.

Thanks for the comments and review! I modified the code as suggested, including adding **@views**, **@inbounds**, renaming the functions, avoiding intermediate array, etc. In addition, I changed the way I deal with rxnarray to make it use as less slicing as possible. I tested the one using slicing versus the one avoiding using slicing using ConstantTPDomain with PME mechanism; the latter one gives us a slightly 1~2 sec speed up.

All above changes and optimization gave us some speed up compared to previously.
![image](https://user-images.githubusercontent.com/45482070/90688703-d3f22b80-e23c-11ea-83e7-48f5fc49b29d.png)

mjohnson541 commented on src/Domain.jl:
```diff
@@ -958,10 +958,10 @@ end
     V = d.V(t)
     ns = y[d.indexes[1]:d.indexes[2]]
     T = y[d.indexes[3]]
-    N = sum(ns)
```
Lets squash all of these "Use thermo variables to calculate N in calcthermo" commits

mjohnson541 commented on src/Domain.jl:
```diff
@@ -797,7 +797,7 @@ end
         diffs = Array{Float64,1}()
     end
     kfs,krevs = getkfkrevs(phase=d.phase,T=T,P=P,C=C,N=N,ns=ns,Gs=Gs,diffs=diffs,V=d.V)
-    return ns,cs,T,P,d.V,C,N,0.0,kfs,krevs,Array{Float64,1}(),Us,Gs,diffs,Cvave
```
Lets squash these "Add output cpdivR to calcthermo" commits

mjohnson541 commented on src/Domain.jl:
```diff
@@ -165,11 +165,12 @@ function ConstantVDomain(;phase::Z,initialconds::Dict{X,E},constantspecies::Arra
 end
 export ConstantVDomain
 
-@with_kw struct ConstantPDomain{N<:AbstractPhase,S<:Integer,W<:Real,W2<:Real,Q<:AbstractArray} <: AbstractVariableKDomain
```
Squash the Add efficiencyinds commits

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1585,6 +1585,16 @@ end
 end
 export calcdomainderivatives!
 
+@inline function jacobianyefficiencyderiv!(jac::S,domain::Union{ConstantTPDomain,ConstantPDomain,ParametrizedTPDomain,ParametrizedPDomain},kinetics::AbstractFalloffRate,efficiencies::Dict{Int64,Float64},rxnarray::Array{Int64,2},rxnind::Int64,Vind::Int64,cs::Array{Float64,1},T::Float64,V::Float64,C::Float64,Ceff::Float64,Kc::Float64) where {S<:AbstractArray}
```
Can you squash the two non-kinetics object specific jacobian efficiencies commits here and then put the three related to the kinetics objects together in a second commit?

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1603,6 +1603,18 @@ end
     end
 end
 
+@inline function jacobianynsderiv!(jac::S,domain::Union{ConstantTPDomain,ParametrizedTPDomain},rxnarray::Array{Int64,2},efficiencyinds::Array{I,1},cs::Array{Float64,1},kfs::Array{Float64,1},krevs::Array{Float64,1},T::Float64,V::Float64,C::Float64) where {S<:AbstractArray,I<:Integer}
```
Can you squash the "Add helper function jacobianynsderiv! that deals with ns partials for" commits?

mjohnson541 commented on src/Domain.jl:
```diff
@@ -1936,3 +1936,11 @@ function getthermovariableindex(domain::Union{ConstantVDomain,ParametrizedVDomai
+
+function getthermovariableindex(domain::Union{ConstantPDomain,ParametrizedPDomain},target::String)
+    if target == "T"
+        return domain.indexes[3]
```
can you squash the "Add getthermovariableindex" commits?
